src/jit/valuenum.cpp

```diff
@@ -6755,54 +6780,73 @@ void Compiler::fgValueNumberTree(GenTree* tree)
-                        //         initialize it, so in this very rare case, we fabricate a value number.
-                        // Case 2) Local variables that represent structs which are assigned using CpBlk.
-                        GenTree* nextNode = lcl->gtNext;
-                        assert((nextNode->gtOper == GT_ADDR && nextNode->gtOp.gtOp1 == lcl) ||
-                               varTypeIsStruct(lcl->TypeGet()));
-                        lcl->gtVNPair.SetBoth(vnStore->VNForExpr(compCurBB, lcl->TypeGet()));
+                        // Temporary, to make progress.
+                        // TODO-CQ: This should become an assert again...
+                        if (lcl->gtVNPair.GetLiberal() == ValueNumStore::NoVN)
+                        {
+                            assert(lcl->gtVNPair.GetConservative() == ValueNumStore::NoVN);
+
+                            // We don't want to fabricate arbitrary value numbers to things we can't reason about.
+                            // So far, we know about two of these cases:
+                            // Case 1) We have a local var who has never been defined but it's seen as a use.
+                            //         This is the case of storeIndir(addr(lclvar)) = expr.  In this case since we only
+                            //         take the address of the variable, this doesn't mean it's a use nor we have to
+                            //         initialize it, so in this very rare case, we fabricate a value number.
+                            // Case 2) Local variables that represent structs which are assigned using CpBlk.
+                            GenTree* nextNode = lcl->gtNext;
+                            assert((nextNode->gtOper == GT_ADDR && nextNode->gtOp.gtOp1 == lcl) ||
+                                   varTypeIsStruct(lcl->TypeGet()));
+                            generateUniqueVN = true;
+                        }
                     }
-                    assert(lcl->gtVNPair.BothDefined());
-                }
 
-                // TODO-Review: For the short term, we have a workaround for copyblk/initblk.  Those that use
-                // addrSpillTemp will have a statement like "addrSpillTemp = addr(local)."  If we previously decided
-                // that this block operation defines the local, we will have labeled the "local" node as a DEF
-                // This flag propagates to the "local" on the RHS.  So we'll assume that this is correct,
-                // and treat it as a def (to a new, unique VN).
+                    if (!generateUniqueVN && (zeroOffsetFldSeq != nullptr))
+                    {
+                        ValueNum addrExtended = vnStore->ExtendPtrVN(lcl, zeroOffsetFldSeq);
+                        if (addrExtended != ValueNumStore::NoVN)
+                        {
+                            lcl->gtVNPair.SetBoth(addrExtended);
+                        }
```

Comment:
If `addrExtended `was `NoVN`, is it correct that we neither have `generateUniqueVN` set to true, nor have we set `lcl->gtVNPair`?

Reply:
I orginally did try that out (setting generateUniqueVN)  but it turs out that doing that caused a few codegen regressions.

Reply:
So - what value number will `lcl` get in that case?

Reply:
THe code above dealing with `wholeLclVarVNP` will either assign 
` lcl->gtVNPair = wholeLclVarVNP;`
or
`lcl->gtVNPair = partialLclVarVNP;`
or 
set `generateUniqueVN = true;`

For the code regression case we had `(typSize == varSize)` so we assigned `wholeLclVarVNP`
and allowe the address of  a struct byref to be CSE-ed with the address of a  (struct byref + zero-offset field)